File: ja2/src/emu/Apple2.java
package emu;

import java.io.IOException;
import paddle.PaddleButtonStates;
import paddle.Paddles;
import paddle.PaddlesInterface;
import speaker.SpeakerClicker;
import video.PictureGenerator;
import video.TextCharacters;
import video.Video;
import video.VideoDisplayDevice;
import video.VideoMode;
import keyboard.HyperMode;
import keyboard.Keyboard;
import keyboard.KeyboardBufferMode;
import keyboard.KeyboardInterface;
import keyboard.KeypressQueue;
import chipset.AddressBus;
import chipset.Memory;
import chipset.PowerUpReset;
import chipset.Slots;
import chipset.Timable;
import chipset.cpu.CPU6502;

/*
 * Created on Jan 28, 2008
 */
public class Apple2 implements Timable
{
	final Memory ram;
	final Memory rom;
	final Slots slots;
	private final VideoMode videoMode;
	private final KeyboardInterface keyboard;
	private final PaddlesInterface paddles;
	private final SpeakerClicker speaker;
	private final AddressBus addressBus;
	private final PictureGenerator picgen;
	private final TextCharacters textRows;
	private final Video video;
	private final CPU6502 cpu;
	private final PowerUpReset powerUpReset;

	public Apple2(final KeypressQueue keypresses, final PaddleButtonStates paddleButtonStates, final VideoDisplayDevice tv, final HyperMode hyper, final KeyboardBufferMode buffered) throws IOException
	{
		this.slots = new Slots();
		this.videoMode = new VideoMode();
		this.keyboard = new Keyboard(keypresses,hyper,buffered);
		this.paddles = new Paddles();
		this.speaker = new SpeakerClicker();
		this.ram = new Memory(AddressBus.MOTHERBOARD_RAM_SIZ);
		this.rom = new Memory(AddressBus.MOTHERBOARD_ROM_SIZ);
		this.addressBus = new AddressBus(this.ram,this.rom,this.keyboard,this.videoMode,this.paddles,paddleButtonStates,this.speaker,this.slots);
		this.picgen = new PictureGenerator(this.videoMode,tv);
		this.textRows = new TextCharacters();
		this.video = new Video(this.videoMode,this.addressBus,this.picgen,this.textRows);
		this.cpu = new CPU6502(this.addressBus);
		this.powerUpReset = new PowerUpReset(this);
	}

	public void tick()
	{
		this.cpu.tick();
		this.video.tick();
		this.paddles.tick();
		this.speaker.tick();
		this.powerUpReset.tick();
	}

	public void powerOn()
	{
		this.ram.powerOn();
		this.cpu.powerOn();
		this.videoMode.powerOn();
		this.video.powerOn();
		this.picgen.powerOn();
		this.powerUpReset.powerOn();
		// TODO clear up all other things for Apple ][ power-on
	}

	public void powerOff()
	{
		this.ram.powerOff();
	}

	public void reset()
	{
		this.cpu.reset();
		this.slots.reset();
	}
}

File: apple2/emu/epple2/OBSOLETE/epple2_old_version/src/apple2.cpp
/***************************************************************************
 *   Copyright (C) 2008 by Chris Mosher,,,   *
 *   chris@mosher.mine.nu   *
 *                                                                         *
 *   This program is free software; you can redistribute it and/or modify  *
 *   it under the terms of the GNU General Public License as published by  *
 *   the Free Software Foundation; either version 2 of the License, or     *
 *   (at your option) any later version.                                   *
 *                                                                         *
 *   This program is distributed in the hope that it will be useful,       *
 *   but WITHOUT ANY WARRANTY; without even the implied warranty of        *
 *   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the         *
 *   GNU General Public License for more details.                          *
 *                                                                         *
 *   You should have received a copy of the GNU General Public License     *
 *   along with this program; if not, write to the                         *
 *   Free Software Foundation, Inc.,                                       *
 *   59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.             *
 ***************************************************************************/
#include "apple2.h"
#include "slots.h"
#include "videomode.h"
#include "keyboard.h"
#include "addressbus.h"
#include "memory.h"
#include "picturegenerator.h"
#include "textcharacters.h"
#include "video.h"
#include "cpu.h"
#include "paddles.h"
#include "paddlebuttonstates.h"
#include "speakerclicker.h"
#include "analogtv.h"
#include "powerupreset.h"
#include "diskcontroller.h"
#include "languagecard.h"

#include <iostream>
#include <fstream>

Apple2::Apple2(KeypressQueue& keypresses, PaddleButtonStates& paddleButtonStates, AnalogTV& tv, HyperMode& fhyper, KeyboardBufferMode& buffered):
	kbd(keypresses,fhyper,buffered),
	ram(AddressBus::MOTHERBOARD_RAM_SIZ),
	rom(AddressBus::MOTHERBOARD_ROM_SIZ),
	addressBus(ram,rom,kbd,videoMode,paddles,paddleButtonStates,speaker,slts),
	picgen(tv,videoMode,this->revision),
	video(videoMode,addressBus,picgen,textRows),
	cpu(addressBus),
	powerUpReset(*this),
	revision(1)
{
}

Apple2::~Apple2()
{
}


void Apple2::tick()
{
	this->cpu.tick();
	this->video.tick();
	this->paddles.tick();
	this->speaker.tick();

	if (this->revision > 0)
		this->powerUpReset.tick();
}

void Apple2::powerOn()
{
	this->ram.powerOn();
	this->cpu.powerOn();
	this->videoMode.powerOn();
	this->video.powerOn();
	this->picgen.powerOn();
	this->powerUpReset.powerOn();
	// TODO clear up all other things for Apple ][ power-on
}

void Apple2::powerOff()
{
	this->ram.powerOff();
}

void Apple2::reset()
{
	this->cpu.reset();
	this->slts.reset();
}
